### PR TITLE
[ONNX] Fix the issue of converting empty list to sequence.

### PR DESCRIPTION
**Pull request by fatcat-z**
When we construct an empty list by python list comprehension, we need to avoid converting the node without inputs to onnx::Concat in shape_type_inference.cpp and peephole.cpp because it will create an invalid Concat node which doesn't have inputs.

In addition, update the code to avoid passing a Sequence input to an onnx::Cast node which doesn't accept Sequence data type as an input.

Add tests for the validation as well.

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -8872,6 +8872,39 @@ def forward(self, x):
         self.run_test(M(2, 1), (x,))
         self.run_test(M([-1, 3], [-2, -1]), (x,))
 
+    @skipIfUnsupportedMinOpsetVersion(13)
+    def test_sequence_to_float(self):
+        class M(torch.nn.Module):
+            def forward(self, x):
```
Comment (neginraoof) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:8878]: Is this PR covering empty lists? The test cases are a bit different.

File: torch/onnx/symbolic_helper.py [hunk 2]
```diff
@@ -215,6 +215,10 @@ def _is_tensor(x):
 def _is_tensor_list(x):
     return isinstance(x.type(), torch._C.ListType) and isinstance(x.type().getElementType(), torch._C.TensorType)
 
+def _is_scalar_list(x):
+    return isinstance(x.type(), torch._C.ListType) and \
```
Comment (neginraoof) [inline torch/onnx/symbolic_helper.py:219]: Can you add comments to explain how you're checking types here? Maybe we can make it a bit more clear?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -8872,6 +8872,39 @@ def forward(self, x):
         self.run_test(M(2, 1), (x,))
         self.run_test(M([-1, 3], [-2, -1]), (x,))
 
+    @skipIfUnsupportedMinOpsetVersion(13)
```
Comment (neginraoof) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:8875]: Thanks for adding the tests. Can you clarify where the empty list shows up in the graph?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 4]
```diff
@@ -8872,6 +8872,39 @@ def forward(self, x):
...
+    def test_sequence_to_float(self):
+        class M(torch.nn.Module):
+            def forward(self, x):
+                result = torch.tensor([1.1 for _ in range(x.shape[1])], dtype=torch.float)
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:8879]: If I understand correctly, this converts to graph like
```
%seq = onnx::Loop(...)
%vec = onnx::ConcatFromSequence(%seq, ...)
```

Is `onnx::ConcatFromSequence` further optimized away, that `%seq` optimized as scan output of loop?

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 3 above)
Comment (fatcat-z) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:8875]: The ATEN graph looks like:

graph(%x.1 : Long(10, 5, strides=[5, 1], requires_grad=0, device=cpu)):
  %17 : Double(requires_grad=0, device=cpu) = prim::Constant[value={1.1}]()
  %18 : Long(requires_grad=0, device=cpu) = prim::Constant[value={1}]()
  %19 : Bool(requires_grad=0, device=cpu) = prim::Constant[value={1}]()
  %20 : Long(requires_grad=0, device=cpu) = prim::Constant[value={6}]()
  %6 : NoneType = prim::Constant()
  %21 : Bool(requires_grad=0, device=cpu) = prim::Constant[value={0}]()
  %8 : float[] = prim::ListConstruct()
  %9 : int[] = aten::size(%x.1) # <string>:7:9
  %10 : Long(device=cpu) = aten::__getitem__(%9, %18) 
  %16 : float[] = prim::Loop(%10, %19, %8) 
    block0(%11 : int, %15 : float[]):
      %12 : float[] = aten::append(%15, %17)
      -> (%19, %12)
  %result.1 : Tensor = aten::tensor(%16, %20, %6, %21) 
  %x.7 : Tensor = aten::add(%x.1, %result.1, %18)
  return (%x.7)

We can see %8 will create an empty list before we call size() to get the tensor's shape.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 4 above)
Comment (fatcat-z) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:8879]: An ONNX graph example is:

graph(%x.1 : Long(10, 5, strides=[5, 1], requires_grad=0, device=cpu)):
  %1 : Double(requires_grad=0, device=cpu) = onnx::Constant[value={1.1}]()
  %3 : Bool(requires_grad=0, device=cpu) = onnx::Constant[value={1}]()
  %7 : float[] = prim::ListConstruct()
  %8 : Long(2, strides=[1], device=cpu) = onnx::Shape(%x.1)
  %10 : Long(device=cpu) = onnx::Constant[value={1}]()
  %11 : Long(device=cpu) = onnx::Gather[axis=0](%8, %10) 
  %12 : float[] = onnx::Loop(%11, %3, %7) 
    block0(%13 : Long(device=cpu), %cond : bool, %14 : float[]):
      %15 : float[] = onnx::SequenceInsert(%14, %1) 
      %17 : Bool(requires_grad=0, device=cpu) = onnx::Identity(%3)
      -> (%17, %15)
  %18 : Tensor = onnx::ConcatFromSequence[axis=0, new_axis=1](%12)
  %19 : FloatTensor = onnx::Cast[to=1](%18) 
  %20 : FloatTensor(device=cpu) = onnx::Add(%x.1, %19) 
  return (%20)

I don't see the optimization happened. Is it supposed to happen? If so, probably I could go deep to check why it didn't happen in this case to see if it is a bug.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 5]
```diff
@@ -229,25 +229,27 @@ Value* CloneValueFromListConstruct(
...
-      Node* unsqueezed_node = createONNXUnsqueeze(
-          n_graph.get(), n_graph->return_node(), new_input, 0, opset_version);
-      unsqueezed.emplace_back(unsqueezed_node->output());
-    }
-    Node* concat_node =
-        n_graph->insertNode(n_graph->create(::c10::onnx::Concat, 1));
-    concat_node->i_(attr::axis, 0);
-    for (auto v : unsqueezed) {
-      concat_node->addInput(v);
+    if (!v->node()->inputs().empty()) {
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:232]: Interesting. Is this part of the code duplicate of the peephole pass?
Do we want to keep both of these?

**Review by neginraoof — COMMENTED** [review]
Can you add unit tests?

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp (hunk 5 above)
Comment (fatcat-z) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:232]: Yes. This similar code block exists in both shape_type_inference and onnx_peephole pass.

1. If we keep this only in shape_type_inference, the switch _onnx_shape_inference would be a little tricky. Because in some cases, we might set it to False to stop shape inference work.

2. If we keep this only in onnx/peephole.cpp file, I checked the tests work as expected. But according to the comments in shape_type_inference file below, seems like we want to retire it one day. @BowenBao  Do you have an idea about why we put the same code in shape_type_inference.cpp too?

  // In jit/passes/onnx/peephole.cpp::eraseListConstruct,
  // prim::ListConstruct is converted to onnx::Concat. The conversion should
  // eventually be moved to symbolic. For now, treat this operator as
  // special case, and change from list type to tensor type. The scalar type
  // is preserved. If the elemtype is Int, insert a onnx::Concat node into
  // the graph.

**Comment by fatcat-z** [comment]
> Can you add unit tests?

Thanks for the reminder, Seems like I lost them somehow. Add them back into this PR.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp (hunk 5 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:232]: consider abstracting the same logic into a function for reuse. For current design this cannot be retired yet. conversion of prim::ListConstruct depends on later nodes that consume it. So we cannot convert it in symbolic function upfront. Thus here in shape inference, we still observe prim::ListConstruct as upstream input, thus require special handle.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 6]
```diff
@@ -229,25 +229,27 @@ Value* CloneValueFromListConstruct(
...
-    }
-    Node* concat_node =
-        n_graph->insertNode(n_graph->create(::c10::onnx::Concat, 1));
-    concat_node->i_(attr::axis, 0);
-    for (auto v : unsqueezed) {
-      concat_node->addInput(v);
-    }
-    return concat_node->output();
+    if (!v->node()->inputs().empty()) {
+      auto lc_node = v->node();
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:233]: I don't think the discussion here is resolved. Is it possible to abstract out the logic in both peephole and shape inference passes?

Comment (fatcat-z) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:228]: Makes sense, and we can do this.

**Review by neginraoof — APPROVED** [review]
LGTM. Thanks!
